### PR TITLE
Ensure correct boost singleton destruction order [13288]

### PR DESCRIPTION
**Pull request by MiguelCompany**
The first time an interprocess synchronization object is created by boost, a singleton is instantiated and its destructor is registered with std::atexit(&atexit_work).

This PR ensures that singleton is created before the DomainParticipantFactory instance so that it is destroyed after all the participants have been deleted.

### CODE REVIEW

File: src/cpp/fastdds/domain/DomainParticipantFactory.cpp [hunk 1]
```diff
@@ -101,6 +102,26 @@ DomainParticipantFactory::~DomainParticipantFactory()
...
 DomainParticipantFactory* DomainParticipantFactory::get_instance()
 {
+    /*
+     * The first time an interprocess synchronization object is created by boost, a singleton is instantiated and
+     * its destructor is registered with std::atexit(&atexit_work).
+     *
+     * We need to ensure that the boost singleton is destroyed after the instance of DomainParticipantFactory, to
+     * ensure that the interprocess objects keep working until all the participants are destroyed.
+     *
+     * We achieve this behavior by having an static instance of an auxiliary struct that instantiates a synchornization
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/fastdds/domain/DomainParticipantFactory.cpp:112]:
```suggestion
     * We achieve this behavior by having an static instance of an auxiliary struct that instantiates a synchronization
```

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Only a typo

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
:ok_hand:

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM

**Comment by MiguelCompany** [comment]
@mergifyio backport 2.4.x 2.3.x

**Pull request merged**